django/db/models/query.py

```diff
@@ -809,6 +809,16 @@ def exists(self):
             return self.query.has_results(using=self.db)
         return bool(self._result_cache)
 
+    def contains(self, obj):
+        """
+        Check if object exists in queryset.
+
+        If the QuerySet is already fully cached, check if object is in cache.
+        """
+        if self._result_cache is None:
+            return isinstance(obj, self.model) and self.filter(pk=obj.pk).exists()
```

Comment:
Hmm, I think if it's not an instance of the model I'd expect it to raise an exception, not just ignore it, but I'm not sure.

Reply:
I'm not sure either. If a programmer makes a mistake not handling a ValueError from this method in some GenericRelation-situation, you get a bug that could have been avoided. If the programmer makes a mistake testing for the wrong object, we're actually hiding a bug.

Reply:
I _think_ there's the potential for a slight discrepancy here - if the `_result_cache` is not None (it's an evaluated queryset) and you fall back to doing `__contains__` on the underlying list, you'd be comparing using `Model.__eq__` right? Ignoring the possibility that the user has overridden the method (and thus there's unavoidably nothing you can do), that method also checks the `concrete_model` meta attribute, for reasons I expect are nuanced but important (my guess: proxy models)

Reply:
@kezabelle Good catch. Turns out this causes a problem when checking if object is in a proxy model queryset. `concrete_model` should be used instead of `isinstance()`.I'll rework and include my expanded tests.